analysis/portanalyzer.go

```diff
@@ -0,0 +1,165 @@
+package analysis
+
+import (
+	"log"
+	"math"
+
+	"github.com/sarchlab/akita/v3/sim"
+)
+
+// PortBandwidthLogger is a hook for logging messages as they go across a Port
+type PortBandwidthLogger struct {
+	sim.LogHookBase
+	sim.TimeTeller
+
+	name      string
+	startTime sim.VTimeInSec
+	endTime   sim.VTimeInSec
+	interval  sim.VTimeInSec
+
+	messageCounter      map[sim.VTimeInSec]int64
+	trafficBytesCounter map[sim.VTimeInSec]int64
+
+	logger *log.Logger
+}
+
+// NewPortMsgLogger returns a new PortMsgLogger which will write into the logger
+func NewPortBandwidthLogger(
+	name string,
+	interval sim.VTimeInSec,
+	timeTeller sim.TimeTeller,
+	logger *log.Logger,
+) *PortBandwidthLogger {
+	h := new(PortBandwidthLogger)
+
+	h.name = name
+	h.logger = logger
+	h.TimeTeller = timeTeller
+	h.interval = interval
+	h.startTime = 0
+	h.endTime = h.CurrentTime()
+	h.messageCounter = make(map[sim.VTimeInSec]int64)
+	h.trafficBytesCounter = make(map[sim.VTimeInSec]int64)
+
+	return h
+}
+
+// Func writes the message information into the logger
+func (h *PortBandwidthLogger) Func(ctx sim.HookCtx) {
+	msg, ok := ctx.Item.(sim.Msg)
+	if !ok {
+		return
+	}
+
+	now := h.CurrentTime()
+
+	floorEndTime := sim.VTimeInSec(math.Round(math.Floor(float64(now/h.interval))*float64(h.interval)*1e10) / 1e10)
+	h.messageCount(floorEndTime)
+	h.trafficBytesCount(msg, floorEndTime)
+
+	diff := sim.VTimeInSec(math.Round(float64(floorEndTime-h.startTime)*1e10) / 1e10)
+
+	msgID := msg.Meta().ID
+	if msgID == "END" {
```

Comment:
I use the `ataxit` mode in monitor [https://gitlab.com/akita/akita//blob/addBandwidthLogger/monitoring/monitor.go#L54] to send the last message to each port.

Reply:
Yes. I did not move the monitor code here. Using `atexit` is definitely a better method. So, is there still a need to use the `END` message?

This is something we can discuss. I feel it is weird to use the `monitor` to attach the hook. So I suggest we make the following changes. 

1. Create a `PerfAnalyzer` struct, which includes a `RegisterComponent` method. The method should be similar to this one. https://github.com/sarchlab/akita/blob/port_analysis/analysis/bufferanalyzer.go#L114
2. Instead of using a lot of CSV files, we should use SQLite instead. We are already using SQLite (file-based SQL database) in the tracing library. I find it is not very difficult to use SQLite. If we decide to use SQLite, the `PerfAnalyzer` should create the `SQLite` database file. 
3. The `BufferAnalyzer` works in a way there is a global instance of the analyzer that hooks to all the buffers. But the `PortAnalyzer` works in a way that each port hooks an individual analyzer. It is better to make them uniform, but this is not a big problem. 
4. The `BufferAnalyzer` and the `PortAnalyzer` periodically write data to the database.

Reply:
So to make it a bit more concrete. We can define structs in this way.

* `PerfAnalyzer`
    * `NewPerfAnalyzer(dbFilename string, period sim.VTimeInSec) *PerfAnalyzer`
        This function creates a new `PerfAnalyzer` with configuration parameters. If we need many parameters, consider using a builder with `With` functions.
    * `RegisterComponent(sim.Component)`
        This method adds all the proper hooks to the component.
    * `AddDataEntry(time sim.VTimeInSec, where, what string, value float64)`
        This method adds data to the SQLite database.

* `BufferAnalyzer`
    Modify this analyzer to use the service provided by `PerfAnalyzer`. Consider converting it to a per-buffer hook.
    
* `PortAnalyzer`
    I think the body of the code is there. But we also need to make it work under the umbrella of the `PerfAnalyzer` and use the services.